File: mobile_client/app/src/main/java/wpam/mobile_client/protocol/ResponseFactory.java
package wpam.mobile_client.protocol;

import java.io.Closeable;
import java.util.ArrayList;

public class ResponseFactory implements Closeable
	{
		private final int RESPONSE_DATA_TYPE_POSITION = 5;
		private ResponseBuilder builder;

		public ResponseFactory()
		{
		}
		public final void close()
		{
		}

		public Response createCommand(ArrayList<Integer> commandInBytes)
		{

			ResponseType type = ResponseType.forValue(commandInBytes.get(RESPONSE_DATA_TYPE_POSITION));
			Response response;
			switch (type)
			{
				case AUTHENTICATE:
				{
					this.builder = new AuthenticateBuilder();
					response = builder.create(commandInBytes);
					break;
				}
				case MOTOR_STATUS:
				{
					this.builder = new MotorStatusBuilder();
					response = builder.create(commandInBytes);
				}
				case BLINDS_STATUS_RSP:
				{
					this.builder = new BlindsStatusResponseBuilder();
					response = builder.create(commandInBytes);
					break;
				}
				case ACK:
				{
					this.builder = new AckResponseBuilder();
					response = builder.create(commandInBytes);
					break;
				}
				case GUARD_STATUS:
				{
					this.builder = new GuardStatusBuilder();
					response = builder.create(commandInBytes);
					break;
				}
				case SENSOR_TAG:
				{
					this.builder = new SensorTagSamplesResbuilder();
					response = builder.create(commandInBytes);
					break;
				}
				default:
				{
					throw new IllegalArgumentException("Score is not valid!");
				}
			}
			return response;
		}
	}

File: server/src/protocol/ResponseFactory.cpp
#include "ResponseFactory.h"

#include "DataResponseBuilder.h"
#include "AckResponseBuilder.h"
#include "ErrorBuilder.h"
#include "MotorStatusBuilder.h"
#include "GuardStatusBuilder.h"
#include "BlindsStatusResponseBuilder.h"
#include "AuthenticateBuilder.h"
#include "SensorTagSamplesResBuilder.h"

using namespace std;
using namespace communication;

ResponseFactory::ResponseFactory()
{}

ResponseFactory::~ResponseFactory()
{}

unique_ptr<Response> ResponseFactory::createCommand(const vector<uint8_t> &commandInBytes)
{
    const auto type = static_cast<ResponseType >(commandInBytes[RESPONSE_DATA_TYPE_POSITION]);

    switch (type)
    {
        case ResponseType::DATA :
            builder_ = make_unique<DataResponseBuilder>();
            return move(builder_->create(commandInBytes));
        case ResponseType::ACK:
            builder_ = make_unique<AckResponseBuilder>();
            return move(builder_->create(commandInBytes));
        case ResponseType::ERROR:
            builder_ = make_unique<ErrorBuilder>();
            return move(builder_->create(commandInBytes));
        case ResponseType::MOTOR_STATUS:
            builder_ = make_unique<MotorStatusBuilder>();
            return move(builder_->create(commandInBytes));
		case ResponseType::GUARD_STATUS:
            builder_ = make_unique<GuardStatusBuilder>();
            return move(builder_->create(commandInBytes));
        case ResponseType::BLINDS_STATUS_RSP:
            builder_ = make_unique<BlindsStatusResponseBuilder>();
            return move(builder_->create(commandInBytes));
        case ResponseType::AUTHENTICATE:
            builder_ = make_unique<AuthenticateBuilder>();
            return move(builder_->create(commandInBytes));
        case ResponseType::SENSOR_TAG:
            builder_ = make_unique<SensorTagSamplesResBuilder>();
            return move(builder_->create(commandInBytes));
        default:
            throw invalid_argument("Received command does not register in factory.");
    }
}